StackExchange.Redis/StackExchange/Redis/RedisSubscriber.cs

```diff
@@ -170,33 +171,81 @@ public bool Remove(Action<RedisChannel, RedisValue> value)
+            }
+
+            private Task SubscribeToSingleServer(ConnectionMultiplexer multiplexer, RedisChannel channel, CommandFlags flags, object asyncState, bool internalCall)
             {
                 var cmd = channel.IsPatternBased ? RedisCommand.PSUBSCRIBE : RedisCommand.SUBSCRIBE;
                 var selected = multiplexer.SelectServer(-1, cmd, flags, default(RedisKey));
 
-                if (selected == null || Interlocked.CompareExchange(ref owner, selected, null) != null) return null;
+                lock (owners)
+                {
+                    if (selected == null || owners.Contains(selected)) return null;
+                    owners.Add(selected);
+                }
 
                 var msg = Message.Create(-1, flags, cmd, channel);
-
+                if (internalCall) msg.SetInternalCall();
                 return selected.QueueDirectAsync(msg, ResultProcessor.TrackSubscriptions, asyncState);
             }
 
+            private Task SubscribeToMasters(ConnectionMultiplexer multiplexer, RedisChannel channel, CommandFlags flags, object asyncState, bool internalCall)
+            {
+                List<Task> subscribeTasks = new List<Task>();
+                var cmd = channel.IsPatternBased ? RedisCommand.PSUBSCRIBE : RedisCommand.SUBSCRIBE;
+                var masters = multiplexer.GetServerSnapshot().Where(s => !s.IsSlave && s.EndPoint.Equals(s.ClusterConfiguration.Origin));
+
+                lock (owners)
+                {
+                    foreach (var master in masters)
+                    {
+                        if (owners.Contains(master)) continue;
+                        owners.Add(master);
+                        var msg = Message.Create(-1, flags, cmd, channel);
+                        if (internalCall) msg.SetInternalCall();
+                        subscribeTasks.Add(master.QueueDirectAsync(msg, ResultProcessor.TrackSubscriptions, asyncState));
+                    }
+                }
+
+                return Task.WhenAll(subscribeTasks);
+            }
+
             public Task UnsubscribeFromServer(RedisChannel channel, CommandFlags flags, object asyncState, bool internalCall)
             {
-                var oldOwner = Interlocked.Exchange(ref owner, null);
-                if (oldOwner == null) return null;
+                if (owners.Count == 0) return null;
 
+                List<Task> queuedTasks = new List<Task>();
                 var cmd = channel.IsPatternBased ? RedisCommand.PUNSUBSCRIBE : RedisCommand.UNSUBSCRIBE;
                 var msg = Message.Create(-1, flags, cmd, channel);
                 if (internalCall) msg.SetInternalCall();
-                return oldOwner.QueueDirectAsync(msg, ResultProcessor.TrackSubscriptions, asyncState);
+                foreach (var owner in owners)
+                    queuedTasks.Add(owner.QueueDirectAsync(msg, ResultProcessor.TrackSubscriptions, asyncState));
```

Comment:
`owners` isn't locked here, so this can go boom under load. The previous interlock above prevented this scenario.